chalice/deploy/packager.py

```diff
@@ -208,3 +205,258 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        return deps_dir
+
+
+class Package(object):
+    PYPI_SDIST_EXTS = ['.zip', '.tar.gz']
+
+    def __init__(self, filename):
+        # type: (str) -> None
+        self.dist_type = 'whl' if filename.endswith('whl') else 'sdist'
+        self.filename = filename
+        self.identifier = self._calculate_identifier()
+
+    def __str__(self):
+        # type: () -> str
+        return '%s(%s)' % (self.identifier, self.dist_type)
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self)
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        return self.identifier == other.identifier
+
+    def __hash__(self):
+        # type: () -> int
+        return hash(self.identifier)
+
+    def _calculate_identifier(self):
+        # type: () -> str
+        # Identiifer that can be fed into pip ie "name==version"
+        if self.dist_type == 'whl':
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            name, version = self._parse_sdist_filename(self.filename)
+        return '%s==%s' % (name, version)
+
+    def _parse_sdist_filename(self, filename):
+        # type: (str) -> Tuple[str, str]
+        # Filenames are in the format {name}-{version}.tar.gz OR
+        # {name}-{version}.zip. These are the only two formats supported by
+        # PyPi for source distributions.
+        for ext in self.PYPI_SDIST_EXTS:
+            if filename.endswith(ext):
+                name, version = filename[:-len(ext)].split('-')
+                return name, version
+        raise InvalidSourceDistributionNameError(filename)
+
+
+class PipRunner(object):
+    """Wrapper around pip calls."""
+
+    def __init__(self, pip_module):
+        # type: (Any) -> None
+        self._pip = pip_module
+
+    @contextlib.contextmanager
+    def _consume_stdout_and_stderr(self):
+        # type: () -> Any
+        try:
+            out, err = sys.stdout, sys.stderr
+            temp_out, temp_err = StringIO(), StringIO()
+            sys.stdout, sys.stderr = temp_out, temp_err
+            yield temp_out, temp_err
+        finally:
+            sys.stdout, sys.stderr = out, err
+
+    def _execute(self, command, args):
+        # type: (str, List[str]) -> int
+        """Executes a pip command with the given arguments.
+
+        As an implementation detail this class assumes that pip was imported
+        and uses pip.main. This gives us the correct version of python that
+        the user is intending to deploy for free, but relies on the pip main
+        function not changing, which is very unlikely.
+        """
+        final_command = [command]
+        final_command.extend(args)
+        # Call pip and hide stdout/stderr
+        with self._consume_stdout_and_stderr():
```

Comment:
Out of curiosity, what does the output look like when you don't capture it? If it does not look good, it is worth not including it but we may want to at least log it in case anything important happened for debugging reasons.